File: BrewPOS-Workspace/brewpos-backend/src/routes/gamification.ts
import { Router } from 'express';
import { prisma } from '../db';

const router = Router();

// --- QUESTS ---

// GET /api/gamification/quests - Ambil semua quest aktif
router.get('/quests', async (req, res) => {
  try {
    const quests = await prisma.quest.findMany({
      where: { active: true },
    });
    res.json(quests);
  } catch (error) {
    res.status(500).json({ error: 'Failed to fetch quests' });
  }
});

// POST /api/gamification/quests - Buat quest baru
router.post('/quests', async (req, res) => {
  try {
    const { name, description, targetValue, rewardPoints, rewardXp, type, targetEntityId } = req.body;
    const quest = await prisma.quest.create({
      data: { name, description, targetValue, rewardPoints, rewardXp, type, targetEntityId },
    });
    res.status(201).json(quest);
  } catch (error) {
    res.status(500).json({ error: 'Failed to create quest' });
  }
});

// PUT /api/gamification/quests/:id - Update quest
router.put('/quests/:id', async (req, res) => {
  try {
    const { id } = req.params;
    const { name, description, targetValue, rewardPoints, rewardXp, type, targetEntityId } = req.body;
    const quest = await prisma.quest.update({
      where: { id: parseInt(id) },
      data: { name, description, targetValue, rewardPoints, rewardXp, type, targetEntityId },
    });
    res.json(quest);
  } catch (error) {
    res.status(500).json({ error: 'Failed to update quest' });
  }
});

// DELETE /api/gamification/quests/:id - Soft delete quest
router.delete('/quests/:id', async (req, res) => {
  try {
    const { id } = req.params;
    const quest = await prisma.quest.update({
      where: { id: parseInt(id) },
      data: { active: false },
    });
    res.json({ message: 'Quest deleted (soft)', quest });
  } catch (error) {
    res.status(500).json({ error: 'Failed to delete quest' });
  }
});

// --- BADGES ---

// GET /api/gamification/badges - Ambil semua badge
router.get('/badges', async (req, res) => {
  try {
    const badges = await prisma.badge.findMany();
    res.json(badges);
  } catch (error) {
    res.status(500).json({ error: 'Failed to fetch badges' });
  }
});

// POST /api/gamification/badges - Buat badge baru
router.post('/badges', async (req, res) => {
  try {
    const { name, description, iconUrl, requiredTransactions } = req.body;
    const badge = await prisma.badge.create({
      data: { name, description, iconUrl, requiredTransactions: parseInt(requiredTransactions) || 0 },
    });
    res.status(201).json(badge);
  } catch (error) {
    res.status(500).json({ error: 'Failed to create badge' });
  }
});

// PUT /api/gamification/badges/:id - Update badge
router.put('/badges/:id', async (req, res) => {
  try {
    const { id } = req.params;
    const { name, description, iconUrl, requiredTransactions } = req.body;
    const badge = await prisma.badge.update({
      where: { id: parseInt(id) },
      data: { name, description, iconUrl, requiredTransactions: parseInt(requiredTransactions) || 0 },
    });
    res.json(badge);
  } catch (error) {
    res.status(500).json({ error: 'Failed to update badge' });
  }
});

// DELETE /api/gamification/badges/:id - Delete badge
router.delete('/badges/:id', async (req, res) => {
  try {
    const { id } = req.params;
    await prisma.customerBadge.deleteMany({ where: { badgeId: parseInt(id) } });
    const badge = await prisma.badge.delete({
      where: { id: parseInt(id) },
    });
    res.json({ message: 'Badge deleted', badge });
  } catch (error) {
    res.status(500).json({ error: 'Failed to delete badge' });
  }
});

// --- REWARDS ---

// GET /api/gamification/rewards - Ambil semua reward aktif
router.get('/rewards', async (req, res) => {
  try {
    const rewards = await prisma.reward.findMany({
      where: { active: true },
    });
    res.json(rewards);
  } catch (error) {
    res.status(500).json({ error: 'Failed to fetch rewards' });
  }
});

// POST /api/gamification/rewards - Buat reward baru
router.post('/rewards', async (req, res) => {
  try {
    const { name, description, pointsRequired } = req.body;
    const reward = await prisma.reward.create({
      data: { name, description, pointsRequired },
    });
    res.status(201).json(reward);
  } catch (error) {
    res.status(500).json({ error: 'Failed to create reward' });
  }
});

// PUT /api/gamification/rewards/:id - Update reward
router.put('/rewards/:id', async (req, res) => {
  try {
    const { id } = req.params;
    const { name, description, pointsRequired } = req.body;
    const reward = await prisma.reward.update({
      where: { id: parseInt(id) },
      data: { name, description, pointsRequired },
    });
    res.json(reward);
  } catch (error) {
    res.status(500).json({ error: 'Failed to update reward' });
  }
});

// DELETE /api/gamification/rewards/:id - Soft delete reward
router.delete('/rewards/:id', async (req, res) => {
  try {
    const { id } = req.params;
    const reward = await prisma.reward.update({
      where: { id: parseInt(id) },
      data: { active: false },
    });
    res.json({ message: 'Reward deleted (soft)', reward });
  } catch (error) {
    res.status(500).json({ error: 'Failed to delete reward' });
  }
});

// POST /api/gamification/redeem - Tukar poin dengan reward
router.post('/redeem', async (req, res) => {
  try {
    const { customerId, rewardId } = req.body;

    const customer = await prisma.customer.findUnique({ where: { id: customerId } });
    const reward = await prisma.reward.findUnique({ where: { id: rewardId } });

    if (!customer || !reward) {
      return res.status(404).json({ error: 'Customer or Reward not found' });
    }

    if (!reward.active) {
      return res.status(400).json({ error: 'Reward is not active' });
    }

    if (customer.points < reward.pointsRequired) {
      return res.status(400).json({ error: 'Insufficient points' });
    }

    // Kurangi poin
    const updatedCustomer = await prisma.customer.update({
      where: { id: customerId },
      data: { points: customer.points - reward.pointsRequired },
    });

    res.json({ message: 'Redeem successful', customer: updatedCustomer });
  } catch (error) {
    res.status(500).json({ error: 'Failed to redeem reward' });
  }
});

// POST /api/gamification/spin - Lucky Spin
router.post('/spin', async (req, res) => {
  try {
    const { customerId } = req.body;
    const SPIN_COST = 50;

    const customer = await prisma.customer.findUnique({ where: { id: customerId } });
    if (!customer) {
      return res.status(404).json({ error: 'Customer not found' });
    }

    if (customer.points < SPIN_COST) {
      return res.status(400).json({ error: 'Insufficient points for Lucky Spin' });
    }

    // Deduct points
    let updatedCustomer = await prisma.customer.update({
      where: { id: customerId },
      data: { points: customer.points - SPIN_COST },
    });

    // Randomize Reward (Points / XP)
    const rand = Math.random();
    let rewardName = '';
    let bonusPoints = 0;
    let bonusXp = 0;

    if (rand < 0.1) {
      // 10% chance for Jackpot (200 XP)
      rewardName = 'JACKPOT! +200 XP';
      bonusXp = 200;
    } else if (rand < 0.4) {
      // 30% chance for +100 Points
      rewardName = 'LUCKY! +100 Points';
      bonusPoints = 100;
    } else if (rand < 0.8) {
      // 40% chance for +50 Points (Break Even)
      rewardName = 'Balik Modal! +50 Points';
      bonusPoints = 50;
    } else {
      // 20% chance for Zonk
      rewardName = 'Zonk! Coba lagi besok.';
    }

    if (bonusPoints > 0 || bonusXp > 0) {
      updatedCustomer = await prisma.customer.update({
        where: { id: customerId },
        data: {
          points: updatedCustomer.points + bonusPoints,
          xp: updatedCustomer.xp + bonusXp,
        },
      });
    }

    res.json({
      message: 'Spin completed',
      rewardName,
      bonusPoints,
      bonusXp,
      customer: updatedCustomer,
    });
  } catch (error) {
    res.status(500).json({ error: 'Failed to perform lucky spin' });
  }
});

export default router;

File: BrewPOS-Workspace/brewpos-flask/app/routes/gamification.py
import random
from flask import Blueprint, request, jsonify
from app.extensions import db
from app.models.quest import Quest
from app.models.badge import Badge
from app.models.reward import Reward
from app.models.customer import Customer, CustomerBadge

gamification_bp = Blueprint('gamification', __name__, url_prefix='/api/gamification')

# --- QUESTS ---

@gamification_bp.route('/quests', methods=['GET'])
def get_quests():
    try:
        quests = Quest.query.filter_by(active=True).all()
        return jsonify([q.to_dict() for q in quests])
    except Exception as e:
        return jsonify({'error': 'Failed to fetch quests'}), 500

@gamification_bp.route('/quests', methods=['POST'])
def create_quest():
    data = request.get_json() or {}
    try:
        quest = Quest(
            name=data.get('name'),
            description=data.get('description'),
            targetValue=int(data.get('targetValue', 1)),
            rewardPoints=int(data.get('rewardPoints', 0)),
            rewardXp=int(data.get('rewardXp', 0)),
            type=data.get('type', 'TOTAL_TRANSACTIONS'),
            targetEntityId=int(data.get('targetEntityId')) if data.get('targetEntityId') else None,
            isDaily=bool(data.get('isDaily', False))
        )
        db.session.add(quest)
        db.session.commit()
        return jsonify(quest.to_dict()), 201
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': 'Failed to create quest', 'details': str(e)}), 500

@gamification_bp.route('/quests/<int:id>', methods=['PUT'])
def update_quest(id):
    data = request.get_json() or {}
    try:
        quest = Quest.query.get(id)
        if not quest:
            return jsonify({'error': 'Quest not found'}), 404

        if 'name' in data: quest.name = data['name']
        if 'description' in data: quest.description = data['description']
        if 'targetValue' in data: quest.targetValue = int(data['targetValue'])
        if 'rewardPoints' in data: quest.rewardPoints = int(data['rewardPoints'])
        if 'rewardXp' in data: quest.rewardXp = int(data['rewardXp'])
        if 'type' in data: quest.type = data['type']
        if 'targetEntityId' in data: quest.targetEntityId = int(data['targetEntityId']) if data['targetEntityId'] else None
        if 'isDaily' in data: quest.isDaily = bool(data['isDaily'])

        db.session.commit()
        return jsonify(quest.to_dict())
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': 'Failed to update quest'}), 500

@gamification_bp.route('/quests/<int:id>', methods=['DELETE'])
def delete_quest(id):
    try:
        quest = Quest.query.get(id)
        if not quest:
            return jsonify({'error': 'Quest not found'}), 404

        quest.active = False
        db.session.commit()
        return jsonify({'message': 'Quest deleted (soft)', 'quest': quest.to_dict()})
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': 'Failed to delete quest'}), 500


# --- BADGES ---

@gamification_bp.route('/badges', methods=['GET'])
def get_badges():
    try:
        badges = Badge.query.all()
        return jsonify([b.to_dict() for b in badges])
    except Exception as e:
        return jsonify({'error': 'Failed to fetch badges'}), 500

@gamification_bp.route('/badges', methods=['POST'])
def create_badge():
    data = request.get_json() or {}
    try:
        badge = Badge(
            name=data.get('name'),
            description=data.get('description'),
            iconUrl=data.get('iconUrl'),
            requiredTransactions=int(data.get('requiredTransactions', 0))
        )
        db.session.add(badge)
        db.session.commit()
        return jsonify(badge.to_dict()), 201
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': 'Failed to create badge'}), 500

@gamification_bp.route('/badges/<int:id>', methods=['PUT'])
def update_badge(id):
    data = request.get_json() or {}
    try:
        badge = Badge.query.get(id)
        if not badge:
            return jsonify({'error': 'Badge not found'}), 404

        if 'name' in data: badge.name = data['name']
        if 'description' in data: badge.description = data['description']
        if 'iconUrl' in data: badge.iconUrl = data['iconUrl']
        if 'requiredTransactions' in data: badge.requiredTransactions = int(data['requiredTransactions'])

        db.session.commit()
        return jsonify(badge.to_dict())
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': 'Failed to update badge'}), 500

@gamification_bp.route('/badges/<int:id>', methods=['DELETE'])
def delete_badge(id):
    try:
        badge = Badge.query.get(id)
        if not badge:
            return jsonify({'error': 'Badge not found'}), 404

        CustomerBadge.query.filter_by(badgeId=id).delete()
        db.session.delete(badge)
        db.session.commit()
        return jsonify({'message': 'Badge deleted', 'badge': badge.to_dict()})
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': 'Failed to delete badge'}), 500


# --- REWARDS ---

@gamification_bp.route('/rewards', methods=['GET'])
def get_rewards():
    try:
        rewards = Reward.query.filter_by(active=True).all()
        return jsonify([r.to_dict() for r in rewards])
    except Exception as e:
        return jsonify({'error': 'Failed to fetch rewards'}), 500

@gamification_bp.route('/rewards', methods=['POST'])
def create_reward():
    data = request.get_json() or {}
    try:
        reward = Reward(
            name=data.get('name'),
            description=data.get('description'),
            pointsRequired=int(data.get('pointsRequired', 0)),
            type=data.get('type', 'NORMAL')
        )
        db.session.add(reward)
        db.session.commit()
        return jsonify(reward.to_dict()), 201
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': 'Failed to create reward'}), 500

@gamification_bp.route('/rewards/<int:id>', methods=['PUT'])
def update_reward(id):
    data = request.get_json() or {}
    try:
        reward = Reward.query.get(id)
        if not reward:
            return jsonify({'error': 'Reward not found'}), 404

        if 'name' in data: reward.name = data['name']
        if 'description' in data: reward.description = data['description']
        if 'pointsRequired' in data: reward.pointsRequired = int(data['pointsRequired'])
        if 'type' in data: reward.type = data['type']

        db.session.commit()
        return jsonify(reward.to_dict())
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': 'Failed to update reward'}), 500

@gamification_bp.route('/rewards/<int:id>', methods=['DELETE'])
def delete_reward(id):
    try:
        reward = Reward.query.get(id)
        if not reward:
            return jsonify({'error': 'Reward not found'}), 404

        reward.active = False
        db.session.commit()
        return jsonify({'message': 'Reward deleted (soft)', 'reward': reward.to_dict()})
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': 'Failed to delete reward'}), 500


# --- REDEEM & SPIN ---

@gamification_bp.route('/redeem', methods=['POST'])
def redeem_reward():
    data = request.get_json() or {}
    customer_id = data.get('customerId')
    reward_id = data.get('rewardId')

    try:
        customer = Customer.query.get(customer_id)
        reward = Reward.query.get(reward_id)

        if not customer or not reward:
            return jsonify({'error': 'Customer or Reward not found'}), 404
        if not reward.active:
            return jsonify({'error': 'Reward is not active'}), 400
        if customer.points < reward.pointsRequired:
            return jsonify({'error': 'Insufficient points'}), 400

        customer.points -= reward.pointsRequired
        db.session.commit()

        return jsonify({'message': 'Redeem successful', 'customer': customer.to_dict()})
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': 'Failed to redeem reward'}), 500


@gamification_bp.route('/spin', methods=['POST'])
def lucky_spin():
    data = request.get_json() or {}
    customer_id = data.get('customerId')
    SPIN_COST = 50

    try:
        customer = Customer.query.get(customer_id)
        if not customer:
            return jsonify({'error': 'Customer not found'}), 404
        if customer.points < SPIN_COST:
            return jsonify({'error': 'Insufficient points for Lucky Spin'}), 400

        customer.points -= SPIN_COST

        rand = random.random()
        reward_name = ''
        bonus_points = 0
        bonus_xp = 0

        if rand < 0.10:
            reward_name = 'JACKPOT! +200 XP'
            bonus_xp = 200
        elif rand < 0.40:
            reward_name = 'LUCKY! +100 Points'
            bonus_points = 100
        elif rand < 0.80:
            reward_name = 'Balik Modal! +50 Points'
            bonus_points = 50
        else:
            reward_name = 'Zonk! Coba lagi besok.'

        if bonus_points > 0:
            customer.points += bonus_points
        if bonus_xp > 0:
            customer.xp += bonus_xp

        db.session.commit()

        return jsonify({
            'message': 'Spin completed',
            'rewardName': reward_name,
            'bonusPoints': bonus_points,
            'bonusXp': bonus_xp,
            'customer': customer.to_dict()
        })
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': 'Failed to perform lucky spin'}), 500
